amqp/types.py

```diff
@@ -0,0 +1,566 @@
+"""Abstract types."""
+import abc
+import asyncio
+import socket
+from array import array
+from datetime import datetime
+from typing import (
+    Any, Awaitable, Callable, IO, List, Mapping,
+    MutableMapping, NamedTuple, Optional, Sequence, SupportsInt,
+    TypeVar, Tuple, Union,
+)
+from .protocol import queue_declare_ok_t
+from .spec import method_sig_t
+from vine import Thenable
+
+Fd = TypeVar('Fd', int, IO)
+Int = TypeVar('Int', SupportsInt, str)
+
+
+class Frame(NamedTuple):
+    type: int
+    channel: int
+    data: bytes
+
+
+ConnectionBlockedCallbackT = Callable[[str], Optional[Awaitable]]
+ConnectionUnblockedCallbackT = Callable[[], Optional[Awaitable]]
+ConnectionFrameHandlerT = Callable[[Frame], Awaitable]
+ConnectionFrameWriterT = Callable[
+    [int,
+     int,
+     Optional[method_sig_t],
+     Optional[bytes],
+     Optional['MessageT'],
+     Optional[float]],
+    Awaitable,
+]
+
+WaitMethodT = Union[method_sig_t, Sequence[method_sig_t]]
+
+
+class TransportT(metaclass=abc.ABCMeta):
+    """Transport type."""
+
+    rstream: asyncio.StreamReader
+    wstream: asyncio.StreamWriter
+
+    connected: bool = False
+    host: str
+    port: int
+    ssl: Any
+    connect_timeout: float
+    read_timeout: float
+    write_timeout: float
+    socket_settings: Mapping
+    sock: socket.socket
+
+    @abc.abstractmethod
+    async def connect(self) -> None:
+        ...
+
+    @abc.abstractmethod
+    def close(self) -> None:
+        ...
+
+    @abc.abstractmethod
+    async def read_frame(self, timeout: float = None) -> Frame:
+        ...
+
+    @abc.abstractmethod
+    async def write(self, s: bytes, timeout: float = None) -> None:
+        ...
+
+
+class ContentT(metaclass=abc.ABCMeta):
+    """Generic content type."""
+
+    CLASS_ID: int
+    PROPERTIES: Sequence[Tuple[str, str]]
+
+    properties: MutableMapping
+    body_received: int = 0
+    body_size: int = 0
+    ready: bool = False
+
+    frame_method: method_sig_t
+    frame_args: str
+
+    @abc.abstractmethod
+    def _load_properties(
+            self,
+            class_id: int,
+            buf: bytes,
+            offset: int = 0,
+            classes: Mapping = None,
+            unpack_from: Callable = None) -> int:
+        ...
+
+    @abc.abstractmethod
+    def _serialize_properties(self) -> bytes:
+        ...
+
+    @abc.abstractmethod
+    def inbound_header(self, buf: bytes, offset: int = 0) -> int:
+        ...
+
+    @abc.abstractmethod
+    def inbound_body(self, buf: bytes):
+        ...
+
+
+class MessageT(ContentT, metaclass=abc.ABCMeta):
+    """Basic message type."""
+
+    body: bytes
+    children: Any
+    channel: 'ChannelT'
+    delivery_info: Mapping[str, Any]
+
+    content_type: str
+    content_encoding: str
+    application_headers: MutableMapping
+    delivery_mode: int
+    priority: int
+    correlation_id: str
+    reply_to: str
+    expiration: str
+    message_id: str
+    timestamp: datetime
+    type: str
```

Comment:
this could be renamed to msg_type

Reply:
I think it's clear as it is since the type is MessageT.